### PR TITLE
Conjunctive queries of ACSets and the UWD algebra of multispans

### PR DESCRIPTION
**Pull request by epatters**
This PR implements:

1. Bipartite free diagrams, which are the generic diagram shape for (co)limits specified UWDs
2. Limits and colimits of bipartite free diagrams of `Fin(Dom)Functions`:
    - Colimits are easy because we don't need to decompose the diagram
    - Limits are harder because we have to decompose the diagram to avoid an asymptotic complexity explosion
3. UWD algebra of multispans, using 2
4. Conjunctive queries of ACSets, using 3

For now, there are only a few basic tests. I'll follow up later with more tests, as well as benchmarks.

This PR is the culmination of a lot of recent work (#344, #347, #348, #351, #355). But it's not the end of the story on ACSets as relational DBs; it's more like the beginning, providing the minimal viable functionality that we can build on over time.

### CODE REVIEW

**Comment by epatters** [comment]
Owen, I've marked you as a reviewer since you're also working on querying (#345). This is also possibly relevant to limits of ACSets.

File: test/wiring_diagrams/Algebras.jl [hunk 1]
```diff
@@ -29,27 +33,64 @@ k = oapply(seq, Dict(:g => g, :h => h))
...
 # Identity for sequential composition.
 seq_id = @relation (a,a) where (a,) begin end
 k = oapply(seq_id, Dict{Symbol,OpenGraph}(), Dict(:a => FinSet(3)))
 @test apex(k) == Graph(3)
 @test feet(k) == [FinSet(3), FinSet(3)]
 
+# Queries of ACSets
+###################
+
+paths2 = @relation (start=u, stop=w) where (u,v,w) begin
```
Comment (olynch) [inline test/wiring_diagrams/Algebras.jl:61]: Wow, this is basically prolog.

File: src/categorical_algebra/FinSets.jl [hunk 2]
```diff
@@ -454,9 +460,147 @@ struct FinSetFreeDiagramLimit{Ob<:FinSet, Diagram<:FreeDiagram{Ob},
   incl::Incl # Inclusion for the "multi-equalizer" in general formula.
 end
 
+function limit(d::BipartiteFreeDiagram{Ob,Hom}) where
```
Comment (olynch) [inline src/categorical_algebra/FinSets.jl:463]: OK, this is the function that I'm going to use downstream for ACSet limits, correct?

File: src/categorical_algebra/CSets.jl [hunk 3]
```diff
@@ -11,12 +11,62 @@ using Reexport
...
+# FinSets interop
+#################
+
+""" Create `FinFunction` for part or subpart of attributed C-set.
+
+The subpart must be of kind `Ob`. For indexed subparts, the index is included.
+"""
+FinFunction(X::ACSet, name::Symbol) = fin_function(X, Val{name})
+
+@generated function fin_function(X::ACSet{CD,AD,Ts,Idxed},
```
[the remark this answers is not in the document]
Comment (epatters) [reply src/categorical_algebra/CSets.jl:29]: Ah, so you are thinking that for every morphism in the indexing category C, there is a functor from C-Set to Arr(Set), the arrow category of Set?

File: src/categorical_algebra/CSets.jl [hunk 4]
```diff
@@ -208,16 +258,27 @@ end
...
 unpack_diagram(span::Multispan{<:AbstractACSet}) =
-  map(Multispan, finsets(apex(span)), unpack_components(legs(span)))
+  map(Multispan, fin_sets(apex(span)), unpack_components(legs(span)))
 unpack_diagram(cospan::Multicospan{<:AbstractACSet}) =
-  map(Multicospan, finsets(apex(cospan)), unpack_components(legs(cospan)))
+  map(Multicospan, fin_sets(apex(cospan)), unpack_components(legs(cospan)))
 unpack_diagram(para::ParallelMorphisms{<:AbstractACSet}) =
   map(ParallelMorphisms, unpack_components(hom(para)))
 
+function unpack_diagram(diagram::BipartiteFreeDiagram{<:AbstractACSet})
```
[the remark this answers is not in the document]
Comment (epatters) [reply src/categorical_algebra/CSets.jl:269]: OK, nice! I'll do that once #352 is merged.

File: src/categorical_algebra/FreeDiagrams.jl [hunk 5]
```diff
@@ -207,16 +211,75 @@ Base.lastindex(para::ParallelMorphisms) = lastindex(para.homs)
...
-#--------------------
+# Diagrams of flexible shape
+############################
 
-# FIXME: Commutative squares are not free diagrams (they commute!) and so do not
-# belong in this module.
+# Bipartite free diagrams
+#------------------------
 
+@present TheoryBipartiteFreeDiagram <: TheoryUndirectedBipartiteGraph begin
```
[the remark this answers is not in the document]
Comment (epatters) [reply src/categorical_algebra/FreeDiagrams.jl:220]: Yeah, computing a general finite limit efficiently seems hard, but this case is more manageable.

In the docstring I tried to explain that depending on how you look at it, these bipartite graphs are either *un*directed or *uni*directed, but it's still confusing. Do you have any suggestions for the names of the types in `Catlab.Graphs.BipartiteGraphs`? That module is brand new and unreleased, so the best time to change it is now!

File: src/categorical_algebra/FinSets.jl (hunk 2 above)
Comment (epatters) [reply src/categorical_algebra/FinSets.jl:463]: I hope so! Is this class of limits general enough to do products and pullbacks of ACSets? I have not thought through how that would work.

**Pull request merged**